### PR TITLE
Add integration tests for Refreshable Materialized Views (mv4)

### PR DESCRIPTION
**Pull request by fm4v**
This test suite covers https://github.com/ClickHouse/ClickHouse/pull/58934 with tests. Some bugs were found, they are described in comments to this PR.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [x]  Disable merge-commit
- [x]  Disable CI cache

### CODE REVIEW

**Comment by fm4v** [comment]
RMV with `APPEND` and  `AS SELECT` query which doesn't read data from table `SELECT
    now() AS a,
    number AS b
FROM numbers(2)` throws error `There is no query or query context has expired. (THERE_IS_NO_QUERY)`

`test_refreshable_mat_view/test.py::test_simple_append`

```
CREATE MATERIALIZED VIEW default.test_rmv_schedule
REFRESH AFTER 30 SECOND APPEND
(
    `a` DateTime,
    `b` UInt64
)
ENGINE = Memory
AS SELECT
    now() AS a,
    number AS b
FROM numbers(2);

SELECT *
FROM system.view_refreshes \G


Row 1:
──────
database:               default
view:                   test_rmv_schedule
status:                 Scheduled
last_refresh_result:    Error
last_refresh_time:      2024-08-08 22:06:58
last_success_time:      ᴺᵁᴸᴸ
duration_ms:            8
next_refresh_time:      2024-08-08 22:07:28
remaining_dependencies: []
exception:              Code: 393. DB::Exception: There is no query or query context has expired. (THERE_IS_NO_QUERY), Stack trace (when copying this message, always include the lines below):

0. ./build_docker/./src/Common/Exception.cpp:111: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000d7b761b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007b68c0c
2. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000007b8030b
3. ./build_docker/./src/Interpreters/Context.cpp:2620: DB::Context::getQueryContext() const @ 0x000000001113cbc9
4. ./build_docker/./src/Analyzer/Resolve/QueryAnalyzer.cpp:4813: DB::QueryAnalyzer::resolveTableFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&, bool) @ 0x00000000115a8ee1
5. ./build_docker/./src/Analyzer/Resolve/QueryAnalyzer.cpp:5154: DB::QueryAnalyzer::resolveQueryJoinTreeNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&) @ 0x00000000115d4d5c
6. ./build_docker/./src/Analyzer/Resolve/QueryAnalyzer.cpp:5407: DB::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::IdentifierResolveScope&) @ 0x0000000011598aff
7. ./build_docker/./src/Analyzer/Resolve/QueryAnalyzer.cpp:0: DB::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>) @ 0x0000000011596eac
8. ./build_docker/./src/Analyzer/Resolve/QueryAnalysisPass.cpp:0: DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x0000000011596705
9. ./build_docker/./src/Analyzer/QueryTreePassManager.cpp:183: DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x00000000115953b6
10. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:0: DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.1500208036845295065) @ 0x00000000116d031d
11. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:160: DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000116ce2fc
12. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:0: DB::InterpreterSelectQueryAnalyzer::getSampleBlock(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x00000000116d0eb9
13. ./build_docker/./src/Storages/StorageMaterializedView.cpp:0: DB::StorageMaterializedView::prepareRefresh(bool, std::shared_ptr<DB::Context>, std::optional<DB::StorageID>&) const @ 0x000000001226668a
14. ./build_docker/./src/Storages/MaterializedView/RefreshTask.cpp:381: DB::RefreshTask::refreshTask() @ 0x0000000012c6cc7e
15. ./src/Common/Stopwatch.h:72: DB::BackgroundSchedulePool::threadFunction() @ 0x0000000010a19420
16. ./build_docker/./src/Core/BackgroundSchedulePool.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x0000000010a1a4c7
17. ./base/base/../base/wide_integer_impl.h:817: void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000d878349
18. ? @ 0x00007ff335f2eac3
19. ? @ 0x00007ff335fc0850
 (version 24.8.1.1554)
retry:                  0
refresh_count:          0
progress:               nan
elapsed:                0
read_rows:              0
read_bytes:             0
total_rows:             0
total_bytes:            0
written_rows:           0
written_bytes:          0
result_rows:            0
result_bytes:           0

```

As example, reading from table in `AS SELECT` query works correctly:
`test_refreshable_mat_view/test.py::test_simple_append_from_table`

**Comment by fm4v** [comment]
`SYSTEM TEST VIEW test_rmv SET FAKE TIME` doesn't work when `OFFSET` or `RANDOMIZE` is used in the RMV interval, so I can't test scheduling correctness when they are used.

**Comment by fm4v** [comment]
In test `test_refreshable_mat_view/test.py::test_long_query` i'm verifying that `progress` in `system.view_refreshes` works correctly for long running queries. 
It works correctly for query `SELECT toDateTime(1) a, 1 b FROM numbers(1000000000)`, but always `inf` when reading from the table: `SELECT * FROM src1`

**Comment by fm4v** [comment]
Test `test_refreshable_mat_view/test.py::test_query_retry` checks setting `refresh_retries=10` which should retry failed query 10 times and seems that retries doesn't work.
Repro:
```
CREATE TABLE tgt1 (a DateTime, b UInt64) ENGINE = MergeTree ORDER BY tuple();
CREATE MATERIALIZED VIEW
test_rmv
REFRESH
EVERY 1 HOUR
SETTINGS refresh_retries=10
TO tgt1
EMPTY
AS SELECT throwIf(1, '111') a;

SYSTEM TEST VIEW test_rmv SET FAKE TIME {next_refresh_time};
``` 

Retry count isn't changed:
```
database:               default
view:                   test_rmv
status:                 Scheduled
last_refresh_result:    Error
last_refresh_time:      2024-08-09 19:00:00
last_success_time:      ᴺᵁᴸᴸ
duration_ms:            35
next_refresh_time:      2024-08-09 19:00:00
remaining_dependencies: []
exception:              Code: 395. DB::Exception: 111: while executing 'FUNCTION throwIf(1_UInt8 :: 1, '111'_String :: 2) -> throwIf(1_UInt8, '111'_String) UInt8 : 3'. (FUNCTION_THROW_IF_VALUE_IS_NON_ZERO), Stack trace (when copying this message, always include the lines below):

0. ./build_docker/./src/Common/Exception.cpp:111: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000d7b761b
1. DB::Exception::createRuntime(int, String const&) @ 0x000000000cb6c48c
2. DB::(anonymous namespace)::FunctionThrowIf::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const (.4455c76ef1992ddf3aecef7e9eb7f880) @ 0x000000000cb6b216
3. DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000007b8073a
4. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087dd5c7
5. DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087ddede
6. DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087df25d
7. ./contrib/boost/boost/smart_ptr/intrusive_ptr.hpp:117: DB::ExpressionActions::execute(DB::Block&, unsigned long&, bool, bool) const @ 0x000000001120cc19
8. ./build_docker/./src/Processors/Transforms/ExpressionTransform.cpp:0: DB::ExpressionTransform::transform(DB::Chunk&) @ 0x000000001308d756
9. ./contrib/llvm-project/libcxx/include/__utility/swap.h:35: DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) @ 0x000000000dac2a13
10. ./build_docker/./src/Processors/ISimpleTransform.cpp:99: DB::ISimpleTransform::work() @ 0x0000000012e24a12
11. ./build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x0000000012e40107
12. ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000012e34870
13. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::PipelineExecutor::execute(unsigned long, bool) @ 0x0000000012e33d02
14. ./build_docker/./src/Storages/MaterializedView/RefreshTask.cpp:0: DB::RefreshTask::refreshTask() @ 0x0000000012c6d231
15. ./src/Common/Stopwatch.h:72: DB::BackgroundSchedulePool::threadFunction() @ 0x0000000010a19420
16. ./build_docker/./src/Core/BackgroundSchedulePool.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x0000000010a1a4c7
17. ./base/base/../base/wide_integer_impl.h:817: void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000d878349
18. ? @ 0x00007fb95cb55ac3
19. ? @ 0x00007fb95cbe7850
 (version 24.8.1.1554)
retry:                  1
refresh_count:          0
progress:               inf
elapsed:                0
read_rows:              1
read_bytes:             1
total_rows:             0
total_bytes:            0
written_rows:           0
written_bytes:          0
result_rows:            0
result_bytes:           0

```

Only one error in logs:
```
2024.08.09 18:09:32.324213 [ 66 ] {} <Error> RefreshTask: Refresh view default.test_rmv failed (attempt 1/11): Code: 395. DB::Exception: 111: while executing 'FUNCTION throwIf(1_UInt8 :: 1, '111'_String :: 2) -> throwIf(1_UInt8, '111'_String) UInt8 : 3'. (FUNCTION_THROW_IF_VALUE_IS_NON_ZERO), Stack trace (when copying this message, always include the lines below):

0. ./build_docker/./src/Common/Exception.cpp:111: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000d7b761b
1. DB::Exception::createRuntime(int, String const&) @ 0x000000000cb6c48c
2. DB::(anonymous namespace)::FunctionThrowIf::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const (.4455c76ef1992ddf3aecef7e9eb7f880) @ 0x000000000cb6b216
3. DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000007b8073a
4. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087dd5c7
5. DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087ddede
6. DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000087df25d
7. ./contrib/boost/boost/smart_ptr/intrusive_ptr.hpp:117: DB::ExpressionActions::execute(DB::Block&, unsigned long&, bool, bool) const @ 0x000000001120cc19
8. ./build_docker/./src/Processors/Transforms/ExpressionTransform.cpp:0: DB::ExpressionTransform::transform(DB::Chunk&) @ 0x000000001308d756
9. ./contrib/llvm-project/libcxx/include/__utility/swap.h:35: DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) @ 0x000000000dac2a13
10. ./build_docker/./src/Processors/ISimpleTransform.cpp:99: DB::ISimpleTransform::work() @ 0x0000000012e24a12
11. ./build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x0000000012e40107
12. ./build_docker/./src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000012e34870
13. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::PipelineExecutor::execute(unsigned long, bool) @ 0x0000000012e33d02
14. ./build_docker/./src/Storages/MaterializedView/RefreshTask.cpp:0: DB::RefreshTask::refreshTask() @ 0x0000000012c6d231
15. ./src/Common/Stopwatch.h:72: DB::BackgroundSchedulePool::threadFunction() @ 0x0000000010a19420
16. ./build_docker/./src/Core/BackgroundSchedulePool.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x0000000010a1a4c7
17. ./base/base/../base/wide_integer_impl.h:817: void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000d878349
18. ? @ 0x00007fb95cb55ac3
19. ? @ 0x00007fb95cbe7850
 (version 24.8.1.1554)

```

File: tests/integration/test_refreshable_mat_view/test.py [hunk 1]
```diff
@@ -0,0 +1,864 @@
...
+
+@pytest.mark.parametrize("interval", INTERVALS_AFTER + INTERVALS_EVERY)
+@pytest.mark.parametrize(
+    "append",
+    [True, False],
+)
+@pytest.mark.parametrize(
+    "empty",
+    [True, False],
+)
```
Comment (al13n321) [inline tests/integration/test_refreshable_mat_view/test.py:490]: Slow. Consider:
 * Generating `append` at random instead having it as a parameter and doubling the test runtime. It shouldn't affect scheduling anyway. Similarly remove or randomize unimportant parameters in other tests.
 * Doing many (all?) intervals in one go. Because IIUC in CI each test usually runs separately and starts whole new clickhouse servers. So having 2 parameter combinations would be much faster than 70.

File: tests/integration/test_refreshable_mat_view/test.py [hunk 2]
```diff
@@ -0,0 +1,864 @@
...
+    # on_cluster,
+    empty,
+    to_clause,
+):
+    table_clause, to_clause_, tgt = to_clause
+
+    create_sql = CREATE_RMV_TEMPLATE.render(
+        table_name="test_rmv",
+        # if_not_exists=if_not_exists,
+        refresh_interval="EVERY 8 SECOND",
```
Comment (al13n321) [inline tests/integration/test_refreshable_mat_view/test.py:600]: Can also test that the first refresh happens immediately, out-of-schedule (unless EMPTY). Use `refresh every 16 second offset {(now + 8 seconds) % 16 seconds}`, then check that the first 3 refreshes happen ~[0, 8, 24] seconds after view creation. (Maybe don't actually wait for the 24-second mark and just check `next_refresh_time`.)

File: tests/integration/test_refreshable_mat_view/test.py [hunk 3]
```diff
@@ -0,0 +1,864 @@
...
+    rmv4 = get_rmv_info(node, "test_rmv")
+
+    if append:
+        append_expected_rows += 2
+        expect_rows(append_expected_rows, table=tgt)
+    else:
+        expect_rows(2, table=tgt)
+
+    assert rmv4["exception"] is None
+    assert rmv4["refresh_count"] == 3
```
Comment (al13n321) [inline tests/integration/test_refreshable_mat_view/test.py:674]: https://github.com/ClickHouse/ClickHouse/pull/60669 removes `refresh_count` (because it seems confusing that it's only since server startup, in uncoordinated mode).

**Review by al13n321 — DISMISSED** [review]
Hi, thanks for testing, sorry I'm taking forever to respond.

Rebased onto https://github.com/ClickHouse/ClickHouse/pull/60669 and pushed some small fixes. Bigger fixes are needed if this will run in CI, see inline comments (some tests look flaky; too many parameter combinations; scheduling model doesn't match RMV behavior).

---

> RMV with APPEND and AS SELECT query which doesn't read data from table SELECT now() AS a, number AS b FROM numbers(2) throws error There is no query or query context has expired. (THERE_IS_NO_QUERY)

Fixed by https://github.com/ClickHouse/ClickHouse/pull/60669

---

> SYSTEM TEST VIEW test_rmv SET FAKE TIME doesn't work when OFFSET or RANDOMIZE is used in the RMV interval, so I can't test scheduling correctness when they are used.

What were the queries and errors? I tried this and everything worked (with and without https://github.com/ClickHouse/ClickHouse/pull/60669):
```
create materialized view a refresh every 1 minute offset 30 second (x Int64) engine Memory as select number as x from numbers(2);
system test view a set fake time '2025-01-01 00:00:00';
[wait a second]
select * from system.view_refreshes;
drop table a;
create materialized view a refresh every 1 minute randomize for 10 second (x Int64) engine Memory as select number as x from numbers(2);
system test view a set fake time '2025-01-01 00:00:00';
[wait a second]
select * from system.view_refreshes;
select * from a;
```

---

> Test test_refreshable_mat_view/test.py::test_long_query verify that progress in system.view_refreshes works correctly for long running queries.
It works correctly for query SELECT toDateTime(1) a, 1 b FROM numbers(1000000000), but always inf when reading from the table: SELECT * FROM src1

EDIT: Fix in https://github.com/ClickHouse/ClickHouse/pull/69234

---

> Test test_refreshable_mat_view/test.py::test_query_retry checks setting refresh_retries=10 which should retry failed query 10 times and seems that retries doesn't work.
Repro:

That's because the fake time stands still. The first retry gets scheduled for 100ms later, and happens if you advance fake time by a second:
```
[...already advanced fake time twice...]
:) select retry from system.view_refreshes

SELECT retry
FROM system.view_refreshes

Query id: abdce691-15dc-4e67-ac6f-9138964b2a2b

   ┌─retry─┐
1. │     3 │
   └───────┘

1 row in set. Elapsed: 0.016 sec. 

:) SYSTEM TEST VIEW test_rmv SET FAKE TIME '2025-01-01 00:00:03'

SYSTEM TEST VIEW test_rmv SET FAKE TIME '2025-01-01 00:00:03'

Query id: d3013479-e434-4170-9dba-55e529376aed

Ok.

0 rows in set. Elapsed: 0.005 sec. 

:) select retry from system.view_refreshes

SELECT retry
FROM system.view_refreshes

Query id: 84700a7e-2e8d-4853-82fd-498a56c600c6

   ┌─retry─┐
1. │     4 │
   └───────┘

1 row in set. Elapsed: 0.012 sec. 

:) SYSTEM TEST VIEW test_rmv SET FAKE TIME '2025-01-01 00:00:04'

SYSTEM TEST VIEW test_rmv SET FAKE TIME '2025-01-01 00:00:04'

Query id: 3abf1bed-577d-455e-b7be-1ef17b19f4ed

Ok.

0 rows in set. Elapsed: 0.004 sec. 

:) select retry from system.view_refreshes

SELECT retry
FROM system.view_refreshes

Query id: 43e90ec0-88d1-441c-80e9-078364e3416f

   ┌─retry─┐
1. │     5 │
   └───────┘

1 row in set. Elapsed: 0.015 sec. 
```

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit deb5fffb45fc1dcc70a9537f8665a57fde44bbe9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68154/deb5fffb45fc1dcc70a9537f8665a57fde44bbe9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60669/1603b3b126580fbaf6dd2ab461ad392eddf2c145/fast_test.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60669/1603b3b126580fbaf6dd2ab461ad392eddf2c145/stateful_tests__asan_.html">✅ success</a></td></tr>
<tr><td>Stateless tests</td><td>Runs stateless functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60669/1603b3b126580fbaf6dd2ab461ad392eddf2c145/stateless_tests__asan__[1_2].html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60669/1603b3b126580fbaf6dd2ab461ad392eddf2c145/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/integration/helpers/client.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 import logging
 from threading import Timer
 
+import numpy as np
```
[the remark this answers is not in the document]
Comment (fm4v) [reply tests/integration/helpers/client.py:7]: Added missing libs to `tests/integration/README.md`
Working with data is easier with a pandas dataframe than with a dictionary, so I'll stick with it

File: tests/integration/test_refreshable_mat_view/schedule_model.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    for i in range(0, len(parts), 2):
+        grouped_parts.append((parts[i], parts[i + 1]))
+
+    sorted_parts = sorted(
+        grouped_parts, key=lambda x: order.index(x[1]), reverse=reverse
+    )
+    return sorted_parts
+
+
+def get_next_refresh_time(schedule, current_time: datetime, first_week=False):
```
[the remark this answers is not in the document]
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/schedule_model.py]: I've decided to get rid of scheduling model since it takes so much time to make it work the same way. I have leave simple tests for that.

File: tests/integration/test_refreshable_mat_view/test.py [hunk 6]
```diff
@@ -0,0 +1,864 @@
...
+        refresh_interval="EVERY 1 HOUR",
+        to_clause="tgt1",
+        select_query=select_query,
+        with_append=False,
+        empty=True,
+    )
+    node.query(create_sql)
+    rmv = get_rmv_info(node, "test_rmv")
+    node.query(f"SYSTEM TEST VIEW test_rmv SET FAKE TIME '{rmv['next_refresh_time']}'")
+    time.sleep(1)
```
[the remark this answers is not in the document]
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/test.py]: I've removed sleeps wherever it was possible to do so.

File: tests/integration/test_refreshable_mat_view/test.py [hunk 7]
```diff
@@ -0,0 +1,864 @@
...
+    "settings",
+    [
+        {},
+        {
+            "refresh_retries": "10",
+            "refresh_retry_initial_backoff_ms": "10",
+            "refresh_retry_max_backoff_ms": "20",
+        },
+    ],
+)
```
[the remark this answers is not in the document]
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/test.py]: I've reduce number of combinations

File: tests/integration/test_refreshable_mat_view/test.py (hunk 1 above)
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/test.py]: The scheduling model has been removed with this test.

File: tests/integration/test_refreshable_mat_view/test.py [hunk 8]
```diff
@@ -0,0 +1,864 @@
...
+        rmv_info = node.query_with_retry(
+            f"SELECT * FROM system.view_refreshes WHERE view='{table}'",
+            check_callback=(lambda r: r.iloc[0]["status"] == wait_status)
+            if wait_status
+            else (lambda x: True),
+            parse=True,
+        ).to_dict("records")[0]
+
+        # Is it time for python clickhouse-driver?
+        rmv_info["next_refresh_time"] = parse_ch_datetime(rmv_info["next_refresh_time"])
```
[the remark this answers is not in the document]
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/test.py]: Let me check failed tests, i think i've added correct waits

File: tests/integration/test_refreshable_mat_view/test.py (hunk 3 above)
Comment (fm4v) [reply tests/integration/test_refreshable_mat_view/test.py]: Removed it too

**Comment by fm4v** [comment]
Thanks for review @al13n321 
I've made some changes:
- Added replicated database tests
- Removed scheduling model since I can't make it to exactly match to scheduling in RMV in reasonable time, leave simpler tests instead

No bugs have found, feature is BULLETPROOF!
